### PR TITLE
Implement replace-all expression

### PR DESCRIPTION
**Pull request by ZeLonewolf**
Unblocks ZeLonewolf/openstreetmap-americana#680

This PR implements a `replace-all` expression for strings.  This functionality is needed to efficiently support name delimiters as described in the osm-americana ticket "[raise semi-colon limit on lists](ZeLonewolf/openstreetmap-americana#680)".

The expression works as follows:

`['replace-all', 'cats;dogs', ';', ' / ']`

This takes the string `cats;dogs` and replaces the `;` character, replacing it with ` / ` and producing `cats / dogs`.

This is my first maplibre PR, so please help me out if I haven't done anything correctly!

## Launch Checklist

 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [x] Link to related issues.
 - [x] Write tests for all new functionality.
 - [x] Document any changes to public APIs.
 - [x] Add an entry to `CHANGELOG.md` under the `## main` section.

### CODE REVIEW

File: test/integration/expression/tests/replace/test.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "expression": ["replace-all", "cats;dogs;ferrets", ";", " and "],
```
Comment (drwestco) [inline test/integration/expression/tests/replace/test.json:2]: Add additional tests to verify string vs regexp behavior, test error conditions, etc.?

Comment (ZeLonewolf) [reply test/integration/expression/tests/replace/test.json]: It turns out the regexp behavior was more complicated than I realized, so I removed that piece to limit this to basic string-replace functionality only.

Also added additional test inputs to verify various edge cases.

**Comment by HarelM** [comment]
This PR and #2064 will be discussed tomorrow in the TSC meeting, feel free to join and present your point of view.

**Comment by 1ec5** [comment]
Thanks to the committee for hearing out the OSM Americana project’s perspective at the meeting today. To recap:

* This expression operator is a conspicuous gap in an otherwise substantial suite of existing string manipulation operators in the style specification.
* String manipulation particularly useful in runtime styling, where the expression may need to vary based on dynamic user inputs. Yet the same manipulations are also useful to styles based on tilesets authored by different third parties.
* Americana’s current use case is to unique a feature’s names after concatenating its name in the user’s dynamically selected language with its names in the local languages. Other use cases have been raised in connection with mapbox/mapbox-gl-js#4100.
* A built-in operator would eliminate the need for unwieldy, limited, and unperformant workarounds, such as the ones detailed in https://github.com/maplibre/maplibre-gl-js/pull/2064#issuecomment-1417431962, as well as in future situations requiring string replacement.

Some counterpoints raised during the call; please correct me if I’ve misrepresented anything:

* Procedurally, the PR should’ve been preceded by an issue or discussion laying out the case for this feature. The mapbox-gl-js issue did not give MapLibre stakeholders enough opportunity to influence the feature’s design. There is a need for a planning process around the style specification.
* Style authors are expected to have control of the contents of the tiles used by the style; they should generate their own tilesets that hard-code presentational modifications into the feature properties.
* Dynamic client-side features necessitate serving tiles from an API that can perform postprocessing on the fly.
* Style expressions are already too complex. Adding more string manipulation operators would encourage style authors to make them even more complex.
* Measuring the performance of a given expression across the whole map is already difficult enough without implementing additional expression operators. (There was a mention about Mapbox crashes that I didn’t follow.)

There were some other points made about arrays in vector tiles that were more germane to #2064 and nyurik/future-mvt#1.

**Comment by HarelM** [comment]
We are moving the spec to its own repo.
I would recommends moving this PR there.
I would also wait a bit to let the dust settle with the docs generation etc as I've seen that the style-spec docs are not updated when updating the style and I think there's a need to improve on that.
Sorry for the late delay in the response.
Same goes for #2064.

**Comment by ZeLonewolf** [comment]
Closing per https://github.com/maplibre/maplibre-gl-js/pull/2059#issuecomment-1450046533

**Comment by ZeLonewolf** [comment]
Closing per https://github.com/maplibre/maplibre-gl-js/pull/2059#issuecomment-1450046533
